Entropy CLI + directory-based projects

urig opened the pull request:
This pull request implements enhancement https://github.com/entropy-lab/entropy/issues/77 by introducing two new features to Entropy:

1. Entropy is now based on *projects*. A *project* has its own directory where you place your code for running experiments. The directory contains a `.entropy` sub-directory where the *sqlite* database file and the *hdf5* data file are stored. The project directory's name is effectively the project's name.

    Whereas before you may have had:
    ```
    📄 experiment1.db
    📄 experiment1.hdf5
    📄 experiment1.py
    ```
    With the new directory-based projects you will have:
    ```
    📁experiment1    <--- project directory
        📁.entropy
            📄 entropy.db
            📄 entropy.hdf5
        📄 experiment1.py
    ```

    Invoking the `SqlAlchemyDB()` constructor with a path to a `.db` file as before will now raise an `EntropyError`. The error instructs you to upgrade the `.db` file to a directory-based project using the `entropy upgrade` at your leisure - please see further details below.

2. Entropy now has its very own CLI. The CLI currently support two commands:

    - `entropy init <path to project directory>` initializes a new project in the given directory
   
    - `entropy upgrade <path to sqlite db file>` will:
        1. Move the `.db` file (and corresponding `.hdf5` file, if it exists) to a new project directory. The directory name will be the original `.db` file's name.
        2. Upgrade the `.db` file to the latest version of Entropy (if needed)
        3. Migrate experiment *results* and *metadata* from the `.db.` file to will to the `hdf5` (if needed)

urig commented:
Hi @talshani. When we migrate an existing entropy `.db` file into the new directory-based project, we move the file to a new directoyr. We also move the `.hdf5` file if present. *Do we also want to move any .py files (or other files) that reside along side it?* TX

talshani commented:
> Hi @talshani. When we migrate an existing entropy `.db` file into the new directory-based project, we move the file to a new directoyr. We also move the `.hdf5` file if present. _Do we also want to move any .py files (or other files) that reside along side it?_ TX

I think we should not handle the organization of py files at the moment, but this should be up to the user.
Once we have the new project structure in place, we will know more, let's wait for the alpha version

talshani reviewed (changes requested):
Can you create a new PR just with test fixes and another with the CLI only?

talshani commented on entropylab/results_backend/sqlalchemy/__init__.py:
```diff
@@ -1,7 +1,14 @@
+
+
+def init_db(path: str):
+    SqlAlchemyDB(path)
```
I think the init_db logic should be here, and the SqlAlchemyDB should call this function, not the other way around
It will be clearer

urig replied:
Thanks. The `init_db()` logic resides mostly in the `_DbInitializer` class and is quite complex. I'd much rather leave it there because it's an entire class.

The method is just for readability. It might not be intuitive that this is what the SqlAlchemyDB ctor does so this method makes it more explicit in code.

urig commented:
Hi @talshani @qguyk. Asking for one last review of a commit I just added: https://github.com/entropy-lab/entropy/pull/79/commits/4d65fc0ded9dbd3f54e3d3ca557e1b33582aa319

This fixes a bug where the `config.py` file was not included in the `entropylab` package because of its location in the repo root. I brought it down into the `entropylab` folder and fixed references to it. @galwiner will soon test manually.

Also we plan to add to the backlog an integ test that will test the CLI after it's packaged. Thanks. 🙏🏻
